subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionResultEvaluator.java

```diff
@@ -64,7 +64,21 @@ public TestExecutionResultEvaluator(
     }
 
     public boolean hasUnmatchedTests() {
-        return resultCount.get() == 0;
+        if (noTestsSelected()) {
+             return false;
+        } else {
+            return resultCount.get() == 0;
+        }
+    }
+
+    private boolean noTestsSelected() {
+        return internalTestExecutionRequest.getTestExecutionDescriptors().isEmpty()
```

Comment:
🤔 Maybe this should be broken down to named `boolean` values, and combining them at the end?
This is not that understandable nor readable.

Reply:
I've refactored this; PTAL